spec/controllers/shared_examples/shared_spec_for_sync_controller.rb

```diff
@@ -357,6 +280,208 @@ def discard_patient(record)
   end
 end
 
+RSpec.shared_examples "a working sync controller that supports region level sync" do
+  let!(:response_key) {
+    model.to_s.underscore.pluralize
+  }
+  let!(:facility_in_same_block) {
+    create(:facility, state: request_facility.state, block: request_facility.block, facility_group: request_facility_group)
+  }
+
+  let!(:facility_in_other_block) {
+    create(:facility, block: "Other Block", facility_group: request_facility_group)
+  }
+
+  let!(:facility_in_other_group) {
+    create(:facility, facility_group: create(:facility_group))
+  }
+
+  let!(:patient_in_request_facility) { create(:patient, :without_medical_history, registration_facility: request_facility) }
+  let!(:patient_in_same_block) { create(:patient, :without_medical_history, registration_facility: facility_in_same_block) }
+  let!(:patient_assigned_to_block) { create(:patient, :without_medical_history, assigned_facility: facility_in_same_block) }
+  let!(:patient_with_appointment_in_block) {
+    create(:patient, :without_medical_history)
+      .yield_self { |patient| create(:appointment, patient: patient, facility: facility_in_same_block) }
+      .yield_self { |appointment| appointment.patient }
+  }
+  let!(:patient_in_other_block) { create(:patient, :without_medical_history, registration_facility: facility_in_other_block) }
+  let!(:patient_in_other_facility_group) { create(:patient, :without_medical_history, registration_facility: facility_in_other_group) }
```

Comment:
When we have this much dependent state happening via `let!`, it's really hard to track what happens for each individual test and what the actual defendant objects at play are.  Also, adding any new specs for this context will end up w/ all this overhead, even if the new specs don't really need all the state.

I realize there is a lot of setup necessary here to have good tests, and these are critical code paths to be tested. My suggestion is that for future PRs we work towards inlining setup or using plain ruby methods to consolidate setup and then call it normally. We don't need tests to be DRY, in fact they are going to be much more readable and useful for us in the future if we have more things inline.

The Thoughtbot crew have written a bunch on the dangers of let: 
* https://thoughtbot.com/blog/lets-not
* https://thoughtbot.com/blog/my-issues-with-let

I also noticed the [RSpec docs themselves](https://www.rubydoc.info/github/rspec/rspec-core/RSpec%2FCore%2FMemoizedHelpers%2FClassMethods%3Alet) recommend against heavy usage of `let` like this:

> Note: let _can_ enhance readability when used sparingly (1,2, or maybe 3 declarations) in any given example group, but that can quickly degrade with overuse. YMMV.

Please consider this feedback non-blocking, but we should start to refactor away from this sort of `let` overuse in the future.

Reply:
I think @harimohanraj89 and I wrote some of these initially and I'm aware of the dangers of `let`. But for these specs, I think it actually does the opposite when you inline, putting this much setup data in each test makes the plethora of cases unreadable and it's very hard to figure out what the test case is actually doing and/or if there's any difference in the setup data for each of them or not – all this is noise for understanding these cases.

These test cases are really really important, we mess this up and our syncs go for a toss. So we wanted to make them as short and easy to read as possible. And more importantly consistent across the board so that one sync controller doesn't behave differently from the rest.

None of this applies in our case, I think:

![Screenshot 2020-12-08 at 11 42 58 AM](https://user-images.githubusercontent.com/50663/101447078-86641080-394a-11eb-82fe-b225cacf4196.png)

* none of it is alien, the tests don't work if you remove any one
* all this setup is necessary for *all* tests in the block-level sync block
* they are as slow as necessary

We could move this to a function, but that doesn't really solve the problem, it just hides away code from one place to another. The setup complexity remains. I generally prefer in-lining seed data, but 90% of the tests we write are fairly non-critical, unlike this.

In fact this silo-ing of the `let`s was so helpful that it literally paved the way for our manual QA because we translated the seed data nearly 1-to-1, [ref](https://docs.google.com/document/d/1Cflct0Y-44IRUVw_5-NptcnNSX1UgAPBiXqoXHq22io/edit#heading=h.5zi8c8m688p4).

Reply:
> We could move this to a function, but that doesn't really solve the problem, it just hides away code from one place to another. The setup complexity remains.

My point about moving it to a function and calling it from specs was that it would be easier to follow in the future for whoever has to maintain this in two years.  I agree totally that all the setup is necessary and these tests are critical, I just want to be able to understand them quickly later. :)